api/intents_api.py

```diff
@@ -0,0 +1,135 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TypedDict
+
+from flask import Response, render_template
+
+from api import converters
+from chromestatus_openapi.models import PostIntentRequest, MessageResponse
+from framework import basehandlers
+from framework import cloud_tasks_helpers
+from framework import permissions
+from internals import processes
+from internals import stage_helpers
+from internals.core_enums import INTENT_STAGES_BY_STAGE_TYPE
+from internals.core_models import FeatureEntry, Stage
+from internals.review_models import Gate
+from pages.intentpreview import compute_subject_prefix
+
+
+class IntentOptions(TypedDict):
+  subject: str
+  feature_id: int
+  sections_to_show: list[str]
+  intent_stage: int|None
+  default_url: str
+  intent_cc_emails: list[str]
+
+
+
+class IntentsAPI(basehandlers.APIHandler):
+
+  def do_get(self, **kwargs) -> Response|dict|str:
+    """Get the body of a draft intent."""
+    feature_id = int(kwargs['feature_id'])
+    # Check that feature ID is valid.
+    if not feature_id:
+      self.abort(400, msg='No feature specified.')
+    feature: FeatureEntry|None = FeatureEntry.get_by_id(feature_id)
+    if feature is None:
+      self.abort(404, msg=f'Feature {feature_id} not found')
+
+    # Check that stage ID is valid.
+    stage_id = int(kwargs['stage_id'])
+    if not stage_id:
+      self.abort(400, msg='No stage specified')
+    stage: Stage|None = Stage.get_by_id(stage_id)
+    if stage is None:
+      self.abort(404, msg=f'Stage {stage_id} not found')
+
+    # Check that the user has feature edit permissions.
+    redirect_resp = permissions.validate_feature_edit_permission(
+        self, feature_id)
+    if redirect_resp:
+      return redirect_resp
+
+    stage_info = stage_helpers.get_stage_info_for_templates(feature)
+    intent_stage = INTENT_STAGES_BY_STAGE_TYPE[stage.stage_type]
+    default_url = (f'{self.request.scheme}://{self.request.host}'
+                   f'/feature/{feature_id}')
+    template_data = {
+      'feature': converters.feature_entry_to_json_verbose(feature),
+      'stage_info': stage_helpers.get_stage_info_for_templates(feature),
+      'should_render_mstone_table': stage_info['should_render_mstone_table'],
+      'should_render_intents': stage_info['should_render_intents'],
+      'intent_stage': intent_stage,
+      'default_url': default_url,
+    }
+
+    return render_template('blink/intent_to_implement.html', **template_data)
+
+  def do_post(self, **kwargs) -> Response|dict|MessageResponse:
+    """Submit an intent email directly to blink-dev."""
+    feature_id = int(kwargs['feature_id'])
+    # Check that feature ID is valid.
+    if not feature_id:
+      self.abort(400, msg='No feature specified.')
+    feature: FeatureEntry|None = FeatureEntry.get_by_id(feature_id)
```

Comment:
Great idea! The `get_specified_feature()` method only checks if the user has view access however, so that permission check will still need to be maintained in the intents method. I've added a `get_specified_stage()` helper now 🙂